### ISSUE TITLE
Support for Colour Managed Images

### ISSUE THREAD

**Issue by StefJoynson** [openseadragon/openseadragon#1215]
As images are rendered as PNGs they do not have an ICC Colour profile. Are there any plans to change this in future so we can display colour managed images?

**Comment by iangilman** [comment]
No plans currently, but it's an interesting idea! What would be involved in adding that support? We draw the images to canvas; is there anything we need to do at that point?

**Comment by StefJoynson** [comment]
We would need to switch to using a different image format, that supports embedded colour profiles, such as Jpeg. That would be the major change, then it should be quite simple, in theory, to ensure that any colour profile in the original image is copied across to the tiles that are rendered.

**Comment by iangilman** [comment]
Well, you can already use JPEGs; the DZI format supports JPEG tiles, for instance. Maybe that's all you need to do?

**Comment by comrade2sun** [comment]
> "Lack of color management and color interoperability is a longstanding complaint about the canvas API."

May I point you to this proposal, it could be of interest:

<a href="https://github.com/WICG/canvas-color-space/blob/master/CanvasColorSpaceProposal.md">Color managing canvas contents - CanvasColorSpaceProposal.md</a>

**Comment by iangilman** [comment]
@comrade2sun good find!

I see there under "current usage" it says "An image resource drawn to a canvas element will have the same appearance as if it were displayed as the replaced content of an HTML element or used as a CSS style value."  That actually sounds like good news for us… If a JPEG drawn to the browser normally is properly color corrected, then a JPEG drawn to canvas will likewise be properly color corrected.

@StefJoynson Can you confirm that JPEG is doing the right thing in OpenSeadragon?

**Comment by comrade2sun** [comment]
@iangilman
From the  _"libvips and nip2 Development and use of the VIPS image processing system"_ blog: 
<a href="http://libvips.blogspot.de/2013/03/making-deepzoom-zoomify-and-google-maps.html?view=classic">Making DeepZoom, Zoomify and Google Maps image pyramids with vips</a>

#### Writing DeepZoom pyramids
I suppose this is of interest ...

> You can set any format write options you like, see the <a href="http://www.vips.ecs.soton.ac.uk/supported/current/doc/html/libvips/VipsForeignSave.html#vips-jpegsave">API docs for vips_jpegsave()</a> for details.

... even more interesting:

> Optional arguments:
> • **profile : filename of ICC profile to attach**
> 
> Use profile to give the filename of a profile to be embedded in the JPEG. [...]
> 
> If no profile is specified and the VIPS header contains an ICC profile named 
> <a href="http://www.vips.ecs.soton.ac.uk/supported/current/doc/html/libvips/libvips-header.html#VIPS-META-ICC-NAME:CAPS"> VIPS_META_ICC_NAME</a>, the profile from the VIPS header will be attached.

In theory(?) it should be possible to attach an ICC profile to each tile.
In practice(???) honestly I have no idea, if it is as simple as pick up - let's say `"sRGB Color Space Profile.icm" ` from `"C:\Windows\System32\spool\drivers\color"` folder - and dump just in.

I would prefer to leave this 'practical part' :sweat_drops: to someone who is more experienced and **_comfortable_** with the CLI as I do ...

**Comment by iangilman** [comment]
@comrade2sun Looks promising! Hopefully @StefJoynson can run with it :)

**Comment by comrade2sun** [comment]
@iangilman 
It does make a difference with or without the sRGB profile ... if you are able to configure your browsers color managment manually 🛠 
 But first of all, who knows and cares about it?
Just for fun, I do recommend this <a href="http://cameratico.com/tools/web-browser-color-management-test/">Web browser color management test ...</a>
... but even much more this very well-written article by Fábio Pili for a necessary insight:

> **<a href="http://cameratico.com/guides/web-browser-color-management-guide/">The sad state of web browser color management</a>**

Let's see ... still I am curious!

@StefJoynson 
PNG files are tagged by default with sRGB profile unless you strip them with `pngcrush` to reduce filesize.

**Comment by jolyonralph** [comment]
This has become an issue for us on mindat.org - as images uploaded with AdobeRGB profile are showing the incorrect color when converted to PNG DZI stacks (using python DeepZoom ImageCreator.)

Some experimenting today showed that:

If I convert a JPEG which has the AdobeRGB profile to a PNG DZI stack with the above tool, every tile has the sRGB profile incorrectly associated with it.

This is why the color is off when viewing such stacks in OpenSeaDragon.

However, if I go and manually load one of the PNG tiles into Photoshop, choose 'Assign Profile' and re-assign the AdobeRGB profile to that image (this command changes the image profile WITHOUT remapping the image - ie for fixing images with the wrong profile) and then save the PNG over the existing tile, clear caches and reload, that particular tile is rendered correctly (at least in latest Chrome) - meaning even a mix of tiles in different color profiles is rendered correctly by the browser to the browser's color profile within the Canvas.

So, the good news is it appears that OpenSeaDragon is doing everything correctly, and it's the python deepzoom ImageCreator script that is messing things up!

For now I'll probably convert the source image to sRGB before tiling just to be safe.

**Comment by iangilman** [comment]
Good to know! I suppose you could try some of the other tiling options:

http://openseadragon.github.io/examples/creating-zooming-images/

VIPS is pretty good, for instance, but I don't know if they do the profile properly. 

Another option might be to write a script that runs through a folder of images and inserts the proper profile?

**Comment by jolyonralph** [comment]
> Another option might be to write a script that runs through a folder of images and inserts the proper profile?

Yes, this is another option. And may well be the best option. But for now the urgency is to fix it for newly uploaded images and we'll get around to fixing existing incorrectly converted images as soon as we can :)

**Comment by jolyonralph** [comment]
And yes, probably do need to replace the tiling script with one that's actually supported. But again, time needed to get a working solution when I'm meant to be moving servers this week (that didn't happen)

**Comment by iangilman** [comment]
I hear you. Good luck!

**Comment by jcupitt** [comment]
Hello, 

vips `dzsave` strips all metadata from tiles as it writes them. Metadata can be large, and you probably don't want a copy of the same metadata on every tile. Additionally, as you say, things like canvas have rather limited profile support, so the metadata is not even very useful. What about CMYK JPG files, for example? Ouch.

PNG does support embedded profiles, though I don't know if the feature is used much.

I think the best solution is to transform to sRGB as part of pyramid generation. You can do this in libvips with:

```
$ vips icc_transform some_huge_file.jpg my_pyramid.dz /usr/share/color/argyll/ref/sRGB.icm
```

And it'll write a deepzoom pyramid in sRGB space, using any profile embedded in the input file as the source colour space. 

There are some notes on building pyramids with vips here:

http://jcupitt.github.io/libvips/API/current/Making-image-pyramids.md.html

**Comment by jcupitt** [comment]
I meant to say, brace yourself for a series of horrible issues regarding profile support. Users will upload CMYK JPG images which have sRGB profiles embedded in them, or RGB PNGs with CMYK profiles, or CMYK TIFFs with both Adobe98 and sRGB profiles, but no actual CMYK profile.

**Comment by iangilman** [comment]
@jcupitt Thank you for your insights here (and elsewhere)!

**Comment by j-hudecek** [comment]
I have a similar problem - 2MB ICC profile that's just too big to append to each tile. So, I have to apply the ICC on the server at a cost of valuable latency and reduced color fidelity for wide gamut displays.

It would be cool if something like SVG's [color-profile](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/color-profile) existed for canvas so that is could be appended just once and used for all images in canvas.

**Comment by iangilman** [comment]
I'm out of my depth in this area, but if you find something like that, we can look into integrating it into OSD!

**Comment by blue-j** [comment]
sRGB is no longer the best choice for output, as P3 devices proliferate.  Ideally there should be a way for images to be color managed throughout the pipeline.  Most images that are not sRGB are going to be Adobe RGB or Image P3, which have _very_ small profiles.  You can also utilize even more compact profiles provided here:    https://github.com/saucecontrol/Compact-ICC-Profiles

The only problem really is CMYK. Those are the profiles that can be enormous.  One thing to share - PNG does not support CMYK - JPEGs actually do (in the wild and from Adobe's products, not necessarily in the official spec).

So with CMYK it is suggested perhaps to transform the source image to sRGB on the back end and then serve sRGB.  In the case of CMYK, sRGB is not that poor of a choice, as most CMYK gamuts are quite small.  [LCMS](https://github.com/mm2/Little-CMS) can do this transformation.

A CSS implementation like "color-profile" has been deprecated and dropped from the most recent CSS colors W3 draft unfortunately.  It is not widely supported for SVGs either.

But this is very important!  OSD is used by museums and professional imaging people who care about color accuracy.  If OSD can support a fully color managed pipeline at least for RGB images, that would be a fantastic differentiator.

**Comment by jcupitt** [comment]
libvips dzsave has had a `--no-strip` option for about a year now. You can use it like this:

```
vips dzsave big.jpg my-pyramid --no-strip
```

And any profile on `big.jpg` will appear on all the tiles in `my-pyramid_files`.

**Comment by blue-j** [comment]
> libvips dzsave has had a `--no-strip` option for about a year now. You can use it like this:
> 
> ```
> vips dzsave big.jpg my-pyramid --no-strip
> ```
> 
> And any profile on `big.jpg` will appear on all the tiles in `my-pyramid_files`.

I understand that that’s what that switch was made to do, but my chief developer says it doesn’t work and has reported the issue. Do you have a good experience with it?

**Comment by jcupitt** [comment]
Yes, someone opened an issue on this a couple of days ago -- it looks like there's a regression in current stable. I plan to fix it today, if I find a little time.

**Comment by jcupitt** [comment]
... it's fixed, 8.10.3 will have this sorted out. Out in a few days, or you can use HEAD of the 8.10 branch.

**Comment by blue-j** [comment]
You’re very kind to attend to this, thank you so much!!  Is there a way I can donate or send you something in gratitude?

I will push my luck here and ask... how difficult would it be to strip all metadata except the ICC profile block?  I foresee that being a common question for people who want color management but not to publish metadata with every tile of add that small unnecessary weight.

**Comment by jcupitt** [comment]
I would use something like python / ruby / php / etc. to strip the unwanted metadata before calling dzsave. 

There's a "Sponsor" button on the libvips github if you're feeling generous.
